src/BSeries.jl

```diff
@@ -1952,16 +1953,18 @@ function _is_energy_preserving_dense(trees, coefficients)
 
     # We remove repeated columns
     unique!(energy_preserving_basis)
+    # Set the numerical tolerance
+    tolerance = 1e-8
 
     # The components of `energy_preserving_basis` are the columns of the matrix `M`.
     M = reduce(hcat, energy_preserving_basis)
-    rank_M = rank(M)
+    rank_M = rank(M, atol = tolerance)
```

Comment:
If we use a tolerance here, we should discuss whether it makes sense to allow users passing the tolerance as keyword arguments.

Reply:
Yes, the tolerance should definitely be passed in by the user.  We should probably default to zero tolerance.

Reply:
The default setting for `rank` includes a non-zero relative tolerance:

```julia
julia> using LinearAlgebra

help?> rank
search: rank RankDeficientException lowrankupdate lowrankupdate! lowrankdowndate lowrankdowndate! readlink

  rank(A::AbstractMatrix; atol::Real=0, rtol::Real=atol>0 ? 0 : n*ϵ)
  rank(A::AbstractMatrix, rtol::Real)

  Compute the rank of a matrix by counting how many singular values of A have magnitude greater than max(atol,
  rtol*σ₁) where σ₁ is A's largest singular value. atol and rtol are the absolute and relative tolerances,
  respectively. The default relative tolerance is n*ϵ, where n is the size of the smallest dimension of A, and
  ϵ is the eps of the element type of A.

  │ Julia 1.1
  │
  │  The atol and rtol keyword arguments requires at least Julia 1.1. In Julia 1.0 rtol is available as
  │  a positional argument, but this will be deprecated in Julia 2.0.
```

I think it does not really make sense to use vanishing tolerances for the computation of the `rank` of matrices in floating point precision. I don't have a good suggestion but the kind of default used there looks reasonable to me.

Reply:
Yes, you're right -- @Sondar74 and I had discussed it and come to the same conclusion.